catalog/app/containers/Bucket/requests/packageUpload.ts

```diff
@@ -0,0 +1,258 @@
+import * as R from 'ramda'
+import * as React from 'react'
+
+import { JsonValue } from 'components/JsonEditor/constants'
+import * as APIConnector from 'utils/APIConnector'
+import { makeSchemaDefaultsSetter, JsonSchema } from 'utils/json-schema/json-schema'
+import pipeThru from 'utils/pipeThru'
+import * as workflows from 'utils/workflows'
+
+// "CREATE package" - creates package from scratch to new target
+// "UPDATE package" - creates package from source (existing manifest) to the same target
+// "COPY package" - creates package from source (existing manifest) to new target
+// "WRAP package" - creates package (wraps directory) from source (bucket + directory) to new target
+
+interface FileEntry {
+  is_dir: boolean
+  logical_key: string
+  path: string
+}
+
+interface FileUpload {
+  hash: string
+  logical_key: string
+  physical_key: {
+    bucket: string
+    key: string
+    version: string
+  }
+  size: number
+}
+
+interface ManifestBodyBase {
+  message: string
+  meta: JsonValue
+  registry: string
+  workflow?: string | null
+}
+
+interface ManifestBodyCreate extends ManifestBodyBase {
+  contents: FileUpload[]
+  name: string
+}
+
+interface ManifestBodyCopy extends ManifestBodyBase {
+  name: string
+  parent: {
+    top_hash: string
+    registry: string
+    name: string
+  }
+}
+
+interface ManifestBodyWrap extends ManifestBodyBase {
+  dst: {
+    registry: string
+    name: string
+  }
+  entries: FileEntry[]
+}
+
+type ManifestBody = ManifestBodyCreate | ManifestBodyWrap | ManifestBodyCopy
+
+const ENDPOINT_CREATE = '/packages'
+
+const ENDPOINT_UPDATE = '/packages'
+
+const ENDPOINT_COPY = '/packages/promote'
+
+const ENDPOINT_WRAP = '/packages/from-folder'
+
+type Endpoint =
+  | typeof ENDPOINT_CREATE
+  | typeof ENDPOINT_UPDATE
+  | typeof ENDPOINT_COPY
+  | typeof ENDPOINT_WRAP
+
+// TODO: reuse it from some other place, don't remember where I saw it
+interface ManifestHandleTarget {
+  bucket: string
+  name: string
+}
+
+interface ManifestHandleSource extends ManifestHandleTarget {
+  revision: string
+}
+
+interface BasePackageParams {
+  message: string
+  meta: JsonValue
+  workflow: workflows.Workflow
+}
+
+interface CreatePackageParams extends BasePackageParams {
+  contents: FileUpload[]
+  target: {
+    bucket: string
+    name: string
+  }
+}
+
+interface UpdatePackageParams extends BasePackageParams {
+  contents: FileUpload[]
+  source: {
```

Comment:
actually, this is not source, but destination -- it tells where to put the manifest. the source is implicit -- update dialog just loads everything from the source manifest to the `contents` array

Reply:
btw, do you prefer `destination` instead of `target`? Probably, `destination` is more explicit

Reply:
i prefer concise names, so `target` seems better, but `dst` would also do =) doesnt matter that much as long as it's used consistently